README.md

```diff
@@ -1,6 +1,238 @@
+
+```ruby
+desc 'outputs hello to the terminal'
+task :hello do 
+  puts "hello from Rake!"
+end
+```
+
+Now, if we run `rake -T` in the terminal, we should see the following:
+
+```bash
+rake hello       # outputs hello to the terminal
+```
+
+So handy!
+
+### Namespacing Rake Tasks
+
+It is possible to namespace your Rake tasks. What does "namespace" mean? A namespace is really just a way to group or contain something, in this case our Rake tasks. So, we might namespace a series of greeting Rake tasks, like `hello`, above, under the `greeting` heading. 
+
+Let's take a look at namespacing now. Let's say we create another greeting-type Rake task, `hola`:
+
+```ruby
+desc 'outputs hola to the terminal'
+task :hola do 
+  puts "hola de Rake!"
+end
+```
+
+Now, let's namespace both `hello` and `hola` under the `greeting` heading:
+
+```ruby
+desc 'outputs hello to the terminal'
+namespace :greeting do 
+  task :hello do 
+    puts "hello from Rake!"
+  end
+  
+  desc 'outputs hola to the terminal'
+  task :hola do 
+    puts "hola de Rake!"
+  end
+end
+```
+
+Now, to use either of our Rake tasks, we use the following syntax:
+
+```bash
+rake greeting:hola
+hello from Rake!
+
+rake greeting:hola
+hola de Rake!
+```
+
+## Common Rake Tasks
+
+As we move towards developing Sinatra and Rails web applications, you'll begin to use some common Rake tasks that handle the certain database-related jobs. 
+
+### `rake db:migrate`
+
+One common pattern you'll soon become familiar with is the pattern of writing code that creates database tables and then "migrating" that code using a rake task. 
+
+Our `Student` class currently has a `#create_table` method, so let's use that method to build out our own `migrate` Rake task. 
+
+We'll namespace this task under the `db` heading. This namespace will contain a few common database-related tasks.
+
+We'll call this task `migrate`, because it is a convention to say we are "migrating" our database by applying SQL statements that alter that database.
+
+```ruby
+desc 'migrate changes to your database' 
+namespace :db do 
+  task :migrate => :environment do 
+    Student.create_table
+  end
+end
+```
+
+Now, if we run `rake db:migrate`, our database table will be created. 
+
+#### Task Dependency
+
+You might be wondering what is happening with this snippet:
+
+```ruby
+task :migrate => :environment do 
+...
+```
+
+This creates a *task dependency*. Since our `Student.create_table` code would require access to the `config/environment.rb` file (which is where the student class and database are loaded), we need to give our task access to this file. We can do so by defining yet another Rake task that we can tell to run before the `migrate` task is run. 
+
+Let's check out that `environment` task:
+
+```ruby
+task :environment do
+  require_relative './config/environment'
+end
+```
```

Comment:
can you put the task def in with this code block to show they go in one file. It might seem like the task :environment goes in the `environment.rb` file